setup.py

```diff
@@ -21,6 +21,7 @@
     "python-dotenv",
     "tiktoken",
     "pydantic>=1.10,<3",  # could be both V1 and V2
+    "docker",
```

Comment:
Can we remove this? The reason is that there are cases docker is not needed. For example when using autogen entirely in a docker container; or when no code execution is enabled. I'd like to keep the required dependency to minimum.

Reply:
can we set up an opt-out with a switch at execution to exclude? IMO, we should be encouraging users to use docker as the build will be consistent and reliable, leading to a better user/developer experience.

Reply:
My understanding is that adding docker to the requirements simply adds the python client, and this should be pretty lightweight and not conflict with anything (especially if in a venv). 

Getting Docker to connect to the Docker service and actually run code -- well that requires being able to host and run containers. But, you can always set use_docker to False. Given the fundamental nature of code execution in AutoGen and the safety afforded by Docker, I'd vote to not fuss around with optional dependencies. 

If the optional dependency specification could include negative dependencies (eg., optionally remove a default dependency), that would be ok too.

Reply:
These are the only dependencies (https://github.com/docker/docker-py/blob/bd164f928ab82e798e30db455903578d06ba2070/setup.py#L12-L15):
```python
requirements = [
    'requests >= 2.26.0',
    'urllib3 >= 1.26.0',
]
```

They are not problematic as such since they are from 2020 and 2021, but they don't allow for execution using Docker either. I would say they are unnecessary unless you install Docker runtime which is not easy at all. Installing docker python lib on top of it is easy.

Reply:
I guess the question is: if we include docker python package as a required dependency, in the scenario when the user does not have docker launcher installed, is there an easy way to inform the user that they can run it locally -- but also inform them the risk of doing so.

Reply:
Yeah that no-docker is what I was referring to when I mentioned "negative" dependencies. I don't know how to do this, but it would be absolutely wonderful if we could.

Reply:
it is a bit of a hack to do it but is probably worth it, let me push a change and we can go from there

Reply:
ok I've gone back and forward with this all day, and I think the best solution is:

- add docker to setup.py under install_required:
  - yes it doesn't always have to be used but the same can be said for `diskcache`. If I am urging the user to use docker then they shouldn't have to pip install something extra from the get go. Can you imagine a user getting docker up and running and then docker python lib isn't installed? Not a good look.
  - it just installs the python lib which is OS agnostic and is fine even if you don't have actual docker installed and running
- `use_docker` defaults to `True` (if the user is already running all of autogen in docker container, this argument is ignored anyway)
- if docker is not running in the users system, very clear message on how to turn it off (argument or set env variable)
- no difference between setting False/None, it will leave you alone if you set either but we instruct to use False


Average user experience will be:

- `pip install pyautogen`
- try to run a first example (e.g. stock)
- throws with runtime error during agent setup asking user what to do with code execution
- user is forced to consider what to do and if they want they can very easily turn docker off and get on with their day but at the same time they need to stop and think

Reply:
This is the best solution for local development, but not for server deployments. 

In future, I believe we'll have to abstract CodeExecutor functionality and allow more options other than local execution, regardless of the use of Docker or not. We should allow for system architectures where code is not executed on the same machine but remotely using both managed and unmanaged resources. E.g. you could execute Python code using Azure Functions to achieve better scalability because I really don't want code execution of one user to exhaust resources and block other users, or you could execute code on an HPC cluster because I have specific hw needs etc.

However, this should be part of some other discussion and some other PR. This is a step in the right direction and it LGTM.